Add `ls` to RPC server

iknox-fa opened the pull request:
resolves #3311 and #3356

### Description
Adds support for the `list` method in the RPC server
Adds new keys for `unique_id` and `original_file_path` to list task when invoked with json output

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

jtcohen6 commented on core/dbt/contracts/rpc.py:
```diff
@@ -63,6 +63,16 @@ class RPCCompileParameters(RPCParameters):
+    resource_types: Optional[List[str]] = None
+    models: Union[None, str, List[str]] = None
+    exclude: Union[None, str, List[str]] = None
+    select: Optional[List[str]] = None
```
I believe the type of `select` should be `Union[None, str, List[str]]` as well. It should work just the same as `models` and `exclude`.

iknox-fa replied:
Fixed!

jtcohen6 commented:
This is working well for me in local testing!

A few comments:
- As I noted above, could we switch the type of `select` to accept a string as well as a list? It should work exactly the same as `models` and `exclude`
- For parity with the existing `list` task, I expected to see [an error](https://github.com/fishtown-analytics/dbt/blob/4d5d0e2150ff2776f5171eb229c367af839b9a35/core/dbt/task/list.py#L46-L55) if `models` is set alongside `select` or `resource_types`, since these are mutually exclusive arguments. This shouldn't block merging the PR if it would be tricky to implement.
- The p0 resolution to #3356 is all we need for now. Could we open a new issue to capture the p1 requirement,  parametrizing the set of returned keys?

iknox-fa commented:
> This is working well for me in local testing!
> 
> A few comments:
> 
> * As I noted above, could we switch the type of `select` to accept a string as well as a list? It should work exactly the same as `models` and `exclude`
Done!
> * For parity with the existing `list` task, I expected to see [an error](https://github.com/fishtown-analytics/dbt/blob/4d5d0e2150ff2776f5171eb229c367af839b9a35/core/dbt/task/list.py#L46-L55) if `models` is set alongside `select` or `resource_types`, since these are mutually exclusive arguments. This shouldn't block merging the PR if it would be tricky to implement.
Done!
> * The p0 resolution to #3356 is all we need for now. Could we open a new issue to capture the p1 requirement,  parametrizing the set of returned keys?
Done!

jtcohen6 commented on test/integration/047_dbt_ls_test/test_ls.py:
```diff
@@ -94,6 +94,8 @@ def expect_snapshot_output(self):
                     'check_cols': None,
                 },
+                'unique_id': 'snapshot.test.my_snapshot',
+                'original_file_path': 'snapshots/snapshot.sql',
```
Looks like this is raising an error on Windows because paths are different. I think we have [`normalize()`](https://github.com/fishtown-analytics/dbt/blob/0f018ea5dd678d37668c819552328b3254c2f220/test/integration/base.py#L32-L42) + [`Normalized()`](https://github.com/fishtown-analytics/dbt/blob/0f018ea5dd678d37668c819552328b3254c2f220/test/integration/base.py#L45-L56) for this purpose.

iknox-fa replied:
Done!

jtcohen6 reviewed (commented):
- There's one last failing test on Windows, I think because of file paths
- I confirmed that passing both `models` and `select` raises an error (good!), passing both `resource_types` params still doesn't raise the expected error (`"models" and "resource_type" are mutually exclusive arguments`). This isn't the most important thing: we're thinking of retiring the `--models` flag/param soon, anyway (#3210).

Once the failing test is sorted, I'm quite happy with where this is at. Nice work!

iknox-fa commented:
> * There's one last failing test on Windows, I think because of file paths
Fixed!
> * I confirmed that passing both `models` and `select` raises an error (good!), passing both `resource_types` params still doesn't raise the expected error (`"models" and "resource_type" are mutually exclusive arguments`). This isn't the most important thing: we're thinking of retiring the `--models` flag/param soon, anyway (#3210).
Fixed!
> 
> Once the failing test is sorted, I'm quite happy with where this is at. Nice work!

kwigley reviewed (approved):
👍

jtcohen6 reviewed (approved):
lgtm!
